### DISCUSSION TITLE
[Chapter3.1-3.4] 저수준 계층? 메모리라는 사물함에서부터 시작해 보자

### DISCUSSION THREAD

**Discussion by ddongule** [elegant-functional-2023/secrets-of-computer#8] [General]
감상평을 남겨주세요~!!

**Comment by ella-yschoi** [comment]
## 감상평

- 3장은 조금 어려웠어요. 지난 1,2장보다 3장은 더 책 닉값을 하는 것 같은.. 정말 밑바닥을 설명한다..
- 책에서는 최대한 쉽게 설명해 주는데.. 나만 이해못한 느낌.. 친절한 교수님 설명에 그렇지 못한 나의 이해력....
- 그동안 맛보기 스푼으로 떠먹었던 스택, 힙 영역을 훨씬 더 깊게 파고파고들어가는 느낌이었습니다. **끝날듯 끝나지 않는 느낌?**
- 3.4장 마지막 즈음에 필자가 _"여러분에게 메모리 할당자는 매우 단순하다는 잘못된 인상을 주지 않기를 바랍니다"_ 라고 했는데요.
- 작전 성공입니다.. 정말 단순하지 않더군요. 후엥
- 어찌보면 뛰어난 개발자들 덕분에 우리가 그나마 **편하게 추상화된 수준으로 개발**을 하고 있는건 아닐까 생각도 들었어요.

<br/>

## 나누고 싶은 이야기

### 3.1장

- 그동안 당연하게 생각했던 변수, 할당, 메모리 주소, 객체를 더 깊은 수준에서 이해할 수 있었습니다.
- 그리고 처음 메모리 주소 관련 이야기에서 프로그래밍에서 말하는 객체는 12바이트 즉, 2의 12승을 말하는건줄 알았어요.
- 찾아보니 객체란 꼭 12바이트 이상을 말하는 것도 아니고, 객체 구성 요소에 따라 바이트 크기는 달라질 수 있다고 하더라고요.
- 물론 자바의 객체 헤더 크기가 12바이트 이긴 하지만 그 외에 객체 필드는 또 바이트 크기가 12바이트가 꼭 아닐 수도 있고요.
- 그래서 필자가 "사물함에 저장되는 것은 0과 1 뿐이며, 모든 개념은 사실 우리가 해석하기 나름입니다" 라고 하는 게 아닐지?

### 3.2장

- 가상 메모리와 물리 메모리 사이에 **맵핑 관계**가 존재하기에 **프로세스에 균일한 가상의 주소 공간**을 제공하는 것이 가능하군요.
- 굳이 맵핑 관계를 어려운 단어인 _'사상 관계'_ 라고 표현해야만 했을까!

### 3.3장

- 3.3장에서는 퀘스트를 비유로 들어 설명하는데, 저는 오히려 이 설명이 더 혼란을 야기했어요 하하
- 218쪽에서 _"프로세스 스택 영역의 높은 주소가 맨 위에 있고 스택 영역은 낮은 주소 방향으로 커진다"_ 라는 문장에 이해가 어려웠는데요.
- 221쪽의 함수의 기계 명령어를 실행하면 반환 주소가 스택 프레임 하단으로 붙어 커지는 그림을 보니 이해가 되었습니다.
- 정리하자면, **스택 프레임은 스택 영역 내에 구성되어 있기에 함수의 호출 단계가 증가할 때마다 스택 영역이 차지하는 메모리가 늘어난다.**
- 이렇듯 간혹 책의 너무나 추상적인 문장들은 이해가 어려워 몇 번을 읽어야 하더라고요.

### 3.4장

- 힙 영역의 특징은 **힙 영역이 필요한 이유**에서부터 비롯해 이해하면 좋네요.
- 프로그래머가 메모리 할당과 해제를 수동으로 관리할 수 있도록 제공되는 메모리 공간, 즉 동적 메모리 할당을 위한 공간.
- 이를 통해 필요한 만큼 메모리를 할당하고, 필요하지 않을 때 메모리를 해제할 수 있다는 것. 
- 메모리 관리 설명을 할 때 주차 비유 설명은 잘 와닿았던 것 같아요.
- 어쨌든 3.4장의 내용을 한 마디로 정리하자면, **메모리를 효율적으로 관리하는 방법** 이라고 할 수 있을듯 합니다.

<br/>

**Reply by sunhpark42** [reply]
저도 사상..? 하고 돌아가서 아 매핑이구나..
그러고 또 다시 사상..? 하며 읽었읍니다ㅋㅋ

**Reply by ddongule** [reply]
인졍인졍ㅋㅋㅋㅋㅋㅋㅋㅋ 중간중간 사상? ;; 이러면서 읽었는데 처음 사상 나올때 사상(mapping) 이렇게 써있더라고요 읽다가 자꾸 까먹어요

**Comment by thgr8ganzi** [comment]
### 감상평

* 포인터는 하나의 변수에 불과하다 메모리 주소를 더 높은 수준으로 추상화 한것이라는 문구가 좋았습니다. 포인터 이해에 도움이 된것 같습니다.
* 참조(reference)를 포인터와 동일한 개념으로 이해 하고있었는데 책에서 포인터를 추상화 한게 참조라는것을 친절히 알려주어서 좋았습니다.
* 힙영역의 메모리 사용 정보는 마치 OSI 7계층에 패킷, 프레임, 세그먼트 등과 비슷한 구조로 되어 있는것 같아요 동일한 개념을 차용 한것 아닌가 싶습니다. 각각 엔드포인트 호스트를 저장소로 보면 어떻게 보면 각각 호스트가 힙 영역과 같은 저장소가 아닌가 싶은...(억지주장)
* 페이지 테이블, 스택프레임의 반환, 점프, 메모리 배치 전략, 메모리 교체 전략, 단편화 등등 공부하였던 것들이 나와서 재밋었는데 매번 까먹는것 같아요.. 이번에도 역시 까먹을것 같습니다.
* 윤태진님이 나미춘이라는건 기억에 남네요ㅠ

**Reply by ddongule** [reply]
나미춘 ;; ㅋㅋㅋㅋㅋㅋㅋㅋㅋ 개웃겨

**Comment by sa02045** [comment]
감상평

- 결국에 데이터는 메모리에서 0과 1로 표현되고, 모든 것들은 0과 1을 추상화한 것에 불과하다는 것이 인상 깊었습니다.
- 프론트엔드 개발하면서 메모리는 거의 신경쓰지 않고 개발을 많이 하는 경우가 많은데, 결국 컴퓨터 동작의 깊은 이해를 위해서는 메모리에 대한 이해가 필수인 것 같습니다.
    - rust를 찍먹해봤을 때 제일 어려운 부분이 소유권, 메모리였는데요. 이 책 공부하고 다시 rust 공부해보면 이해가 더 잘 가지않을까 희망회로를 돌려봅니다…
- 가상 메모리도 사실 물리 메모리의 추상화로 프로세스들이 물리 메모리에 저장될 때는 중구난방으로 흩어져 있다.
- 함수들이 호출될 때 스택 영역에 쌓이면서 구체적으로 어떻게 호출/반환이 이루어지는지 배웠습니다. 결국 메모리와 메모리 주소에 관한 얘기였습니다. 함수를 호출했다면 어디로 가야하는지? 어떻게 다시 돌아올 수 있을지?
- 함수 호출 스택과 CPU 제어권에 대한 내용을 2장에서 배웠던 동기, 비동기 개념과 같이 생각해볼 수 있어서 좋았습니다.
    - 동기: 함수 B의 실행이 끝날때까지는 함수 A로 반환하지 않겠다!
    - 비동기: 함수 B가 실행되자마자 함수 A로 반환하고, 함수 B는 다른 실행흐름에서 실행된다

**Reply by sunhpark42** [reply]
rust 교수 승희님 기대합니다.

**Reply by ddongule** [reply]
러교승

**Reply by sunhpark42** [reply]
나미춘을 이은 러교승

**Reply by iborymagic** [reply]
별다줄

**Reply by yunseorim1116** [reply]
러-강-기(러스트 강의 기대한단뜻)

**Comment by yunseorim1116** [comment]
# 감상평

포인터랑 참조값이랑 다른거였구나 
많은 코린이들이 포인터에서 c를 포기한다는 소문을 들었는데 첨하면 헷갈리긴 할 것 같다.

옛날에 js 가비지 컬렉터가 어떻게 동작하는지 어느 타이밍에 수거되는지 어디에 저장되고 어디에 사라지는지 확인하고 싶었는데 그때는 js는 메모리 주소를 직접적으로 알수 없다는 사실을 몰라서 메모리 주소를 찍어보려고 이것저것 검색했던 기억이 난다. (빨리 알았더라면…)

이번 파트를 읽고 메모리주소/포인터/참조 에 대한 개념이 잡혔다. 세개 다 똑같은 말인줄 알았는데 사실은 조금씩 다른 개념이었던것.

# 궁금증
1. 레지스터는 CPU의 일부분, 이건 물리메모리일까 논리 메모리일까?
2. 8비트로 만들수 있는 조합은 고작 256개뿐, 그래서 더 많은 숫자를 표현하기 위해 인트는 보통 4바이트다. 그럼 4바이트로 모든 수를 표현할수 있을까? 4바이트가 넘어가는건 없을까?
3. 스택프레임과 실행컨텍스트는 어떤 차이가 있을까
4. 왜 1바이트는 8비트일까?

**Reply by ella-yschoi** [reply]
오 스택 프레임과 실행 컨텍스트 차이 한번 파보고 싶네요

**Reply by ddongule** [reply]
오 덕분에 저도 찾아봤는데 아주 아주 간략히 아래처럼 정리할 수 있을 것 같아요
틀린 정보가 있다면 알려주시옵소서

스택 프레임 - 함수가 호출되는 순간 그 함수의 호출 정보들을 저장하기 위해 할당되는 스택
실행 컨텍스트 - 자바스크립트에서 코드가 실행되기 위해 필요한 정보를 담고있는 객체

`실행 컨텍스트`가 호출되면 `스택 프레임`이라는 객체가 생성이 되어 실행 컨텍스트의 내용이 저장됨

- 실행 컨텍스트는 코드의 실행 환경을 `추상화`한 객체
- 스택 프레임은 실행 컨텍스트를 `구체적`으로 표현한 객체

**Comment by sunhpark42** [comment]
- 오랜만에 메모리 구조... 물리적인 개념이 아니라는 것을 늘 생각하기.
- 3.4에서 메모리 동적할당의 개념이 나오는데 이때 free를 한다고 실제 데이터가 지워지진않는다. 단순 연결을 끊는 것 뿐이라, c언어 정도의 저수준 언어를 사용할때에는 데이터의 초기화는 늘 고려해야하는 부분 (덧으로.. free를 안하면 메모리 누수가 발생하는데, 42서울 한창 할 때 이거 누수나면 바로 과제 fail나고 고치고 다시 3번 검사맡아야 했다. 목차보니까 3장 뒷쪽에 나오는듯) 
- 읽을수록 이거 고안하고 만든사람들.. 세기의 천재가 아닌가 싶다..적어도 3세기 정도는 이사람 들이 만든 기본이 세상을 차지하겠지.. 이미 벌써 1세기는 곧 될듯.

**Comment by iborymagic** [comment]
<감상평>
- 예전에 유튜브에서 모 교수님이 대충 `코드 한 줄 한 줄을 짤 때마다 메모리에서 어떤 일이 발생하는지를 이해하면 어디든 갈 수 있다~` 라는 식의 말을 했었더랬죠..
- 요번 장을 읽으면서 뭔가 거기에 한 발짝 (100발짝 더 가야됨) 다가간 느낌이라 기분 좋게 읽었읍니다.. 뭔가 너무 로우 레이어의 일들은 알아서 아저씨가 잘 생략해주는 느낌이랄까요..
- 예전에 C를 찍먹할 때 포인터는 무서워서 대충 이해하고 뛰어넘었는데, 요 아저씨가 포인터를 참조와 엮어서 설명해줘서 비교적 두려움 없이 잘 읽을 수 있었읍니다..
- 마지막에 힙 영역 할당 전략을 설명해줄 때는 메모리 정렬이 없고 최대 크기가 2GB라는 등 제한을 이것저것 걸길래 흠.. 했는데 실제로도 예전 32bit 환경에서는 간단한 malloc 명령어로 할당할 수 있는 메모리 최대 크기가 2~3GB 정도 됐다고 합니다.. 나머지는 커널 메모리 영역으로 남겨둔다고 합니다.. 물론 출처가 모르는 사람들의 토론을 기반으로 하였기 때문에.. 좀 더 찾아볼 필요는 있어보입니다.. 하지만 지금은 책 읽기에도 급급한 시간이므로.. 일단 패쓰.. (https://www.quora.com/What-happens-when-you-malloc-8GB-of-memory-on-a-32-bit-system-with-only-4GB-RAM)
- 가상 메모리를 사용하는 시스템에서는 실제 물리 메모리보다 훨씬 큰 용량도 할당할 수 있다고 하는데.. 그 방식도 궁금쓰 합니다..

**Reply by hanana1253** [reply]
100발짝 더 가야됨 ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ

**Reply by sunhpark42** [reply]
간단하다.. 하지만 간단하지 않다... 궁금하다면 대학원에 와라(?)

**Reply by ddongule** [reply]
ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ대학원 ;; ㅋㅋㅋㅋㅋㅋ ㅠㅠ

**Comment by ddongule** [comment]
## 감상평

메모리에 대해 대략적으로 이해할 수 있었다.
메모리는 0과 1만 저장(bit)하고, 더 많은 정보를 표현하기 위해 bit를 8개로 묶고, 묶음에 주소를 붙여(메모리 주소) 더 쉽게 정보를 찾을 수 있게 한다.
그치만 이 정보도 표현할 수 있는 내용에 한계가 있기 때문에 여러 개의 바이트를 묶어 어떤 구조체(객체)를 만들고 그 구조체를 통해 정보를 나타낸다.
그리고 모든 건 여기에서 시작된다. (?) ㅋㅋㅋ

읽을수록 이딴걸 누가 설계했지? 그 설계한걸 또 어떻게 만들었지? 도데체 어떻게 했지? 내가 사용하는 컴퓨터에 이게 다 녹아 들어가있는거라고? ;; 세상엔 정말 미췬똑똑이들이 많구나? 나랑 같은 휴먼이 맞나? ㄹㅇ 지구를 발전시키기 위해 잠깐 찾아와줬던 외계인들아닐까? 를 느꼈다 ;; 

아 그리고 중간중간에 글 아래 역주에다가 사실은 아니다 이런 내용의 글들이 많아서 긴장감있게 책을 읽어야했다.(?) 솔직히 다른 책이었으면 이놈의 작가놈이 왜 위아래 말을 다르게하누 하면서 짜증이 났을 것 같은데 이 작가는 참 마음에 들어서 짜증이 나지 않았다 ^^7 ㅎㅎㅎ ㅎㅎㅎㅎ

**Reply by sunhpark42** [reply]
ㅋㅋㅋㅋㅋ 역주 너무 웃겨요..

**Reply by ella-yschoi** [reply]
ㅋㅋㅋㅋㅋㅋㅋㅋ 공감됩니닼ㅋㅋㅋㅋ

**Reply by hanana1253** [reply]
> 세상엔 정말 미췬똑똑이들이 많구나? 나랑 같은 휴먼이 맞나? ㄹㅇ 지구를 발전시키기 위해 잠깐 찾아와줬던 외계인들아닐까? 

ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ 매번 민경님 감상평 기대가 됩니다

**Comment by hanana1253** [comment]
## 3.1 
3년 전.. 자바스크립트 선생님이 프로그래밍 배울때 쉽게 이해시키려고 "변수는 값을 담는다"고 표현하는데 그건 사실은 틀리다. 라고 강조했던 기억이 난다.
대신 변수는 데이터를 저장하는 메커니즘이자, 메모리 주소를 가리키는 것이라고 했는데... 당시에는 잘 못알아들었는데 이제야 조금 공감이 되는군.
왜 포인터를 C언어에서는 일반적인 변수와 달리 표기하는지, 그리고 다른 언어들에선 왜 안 그래도 되는지에 대해 알 수 있었다 (추상화의 차이였다니!)

1바이트가 메모리셀 기본 용량인 이유:
- 최초 컴퓨터의 목적은 군사적인 목적, 암호해독/계산을 위한 단일목적을 가짐
- 범용적인 목적을 가지기 시작하며 다양한 데이터를 다뤄야 하는데, 통일되지 않은 단위로 데이터를 관리하면 어려우니 영문 대소문자와 특수기호 등을 포함한 아스키코드를 기준으로 8 bit를 1 Byte로 데이터 저장의 최소 단위를 삼고, 숫자를 통해 문자를 표현하게끔 했다.
- 이후 다른 언어까지 다룰 수 있도록 유니코드로 인코딩 방식이 바뀌었다.

## 3.2
mapping보다 어려운 사상이라는 말..  처음들어본다..
> [IT용어]하나의 값을 다른 값으로 또는 한 데이터 집합을 다른 데이터 집합으로 번역하거나, 2개의 데이터 집합 사이에 1:1 대응 관계를 설정하는 것. 컴퓨터 그래픽스에서 3차원 대상을 평면에, 또는 2차원 이미지를 구체(球體:sphere) 위로 사상하는 것 등이 있다. 또한 가상 기억 시스템에 기억되어 있는 가상 주소를 컴퓨터가 물리적 주소로 번역하는 것도 사상하는 것이다. 도메인 이름을 인터넷 주소로 번역하는 서버 프로그램의 기능이나 메시지 통신 처리 시스템(MHS)에서 디렉터리 이름을 발신자/수신자 주소(O/R address)로 번역하는 메시지 전송 시스템(MTS) 기능도 사상하는 것이다.

p.213 "페이지 단위로 해야 항목 수가 대폭 줄어든다"를 잘 이해하지 못했다

> 이 중에서 코드 영역과 데이터 영역은 이미 1장에서 설명했으므로... 

벌써 기억이 안나는데 어떡하죠 선생님...

## 3.3 

p.221 비어있는 스택 프레임은 뭘 의미하는 거지? 아 매개변수 같은게 담기는 거구나

p.223  매개변수 수가 레지스터 수보다 많으면 나머지 매개변수는 스택 프레임에 직접 넣을 수 있으므로...? 이렇게 동적으로? 

p.224 
> 책을 주의 깊게 읽어온 독자라면 ... 

ㅋ 미안..

## 3.4
는 못읽어서 호스트 걸리면 이제 같이 읽어야함.. ㅎ 죄송합니다.

[정리 링크](https://teal-line-f3e.notion.site/CH-03-d845a33306c6473fbe585df77f1e194a?pvs=4)

**Reply by yunseorim1116** [reply]
정리 링크 안보여요 !! 보고싶음다 !

**Reply by hanana1253** [reply]
세상에 매번 똑같이 실수하다니... 고쳐놓았읍니다

**Reply by ddongule** [reply]
미안..이라니 개웃겨욬ㅋㅋㅋㅋㅋㅋㅋㅋ  ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ ㅠㅠ

**Reply by sunhpark42** [reply]
숫자와.. 0과 1로 만든 표..
![image](https://github.com/elegant-functional-2023/secrets-of-computer/assets/67677561/d663bcf3-1cfa-4ec5-960f-51e7d6481343)
![image](https://github.com/elegant-functional-2023/secrets-of-computer/assets/67677561/dbb0c41d-9200-4590-8999-ceb1764462c2)
